### PR TITLE
Introduce ExecutedObservation retrieval from an External API

### PR DESCRIPTION
**Pull request by mcoughlin**
This PR introduces ExecutedObservation retrieval from an External API.

@robertdstein Building on this would be a good mix of backend and front-end stuff.

### CODE REVIEW

File: skyportal/facility_apis/ztf.py [hunk 1]
```diff
@@ -407,6 +412,47 @@ def remove(request):
...
+        """
+
+        altdata = allocation.altdata
+        if not altdata:
+            raise ValueError('Missing allocation information.')
+
+        jd_start = Time(start_date, format='datetime').jd
+        jd_end = Time(end_date, format='datetime').jd
+
+        if not jd_start < jd_end:
```
Comment (guynir42) [inline skyportal/facility_apis/ztf.py:433]: Is there any clever reason not to do `jd_start >= jd_end` that has to do with `None` values or something?

File: skyportal/facility_apis/ztf.py [hunk 2]
```diff
@@ -407,6 +412,47 @@ def remove(request):
...
+        s = Session()
+        s.auth = (altdata['tap_username'], altdata['tap_password'])
+        client = pyvo.dal.TAPService(altdata['tap_service'], session=s)
+
+        request_str = """
+            SELECT field,rcid,fid,expid,obsjd,exptime,maglimit,ipac_gid,seeing
+            FROM ztf.ztf_current_meta_sci WHERE (obsjd BETWEEN {0} AND {1})
+        """.format(
+            jd_start, jd_end
+        )
```
Comment (guynir42) [inline skyportal/facility_apis/ztf.py:445]: Could we just use an f-string?

File: skyportal/handlers/api/observation.py [hunk 3]
```diff
@@ -613,3 +616,65 @@ def post(self):
...
+
+        try:
+            instrument.api_class_obsplan.retrieve(
+                allocation, data['start_date'], data['end_date']
+            )
+            self.push_notification(
+                'Observation ingestion in progress. Should be available soon.'
+            )
+        except Exception as e:
+            return self.error(f"Error in querying instrument API: {e}")
```
Comment (guynir42) [inline skyportal/handlers/api/observation.py:680]: I don't get it: why retrieve at the end of the post?

File: static/js/components/NewAPIObservation.jsx [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    return errors;
+  }
+
+  const allocationFormSchema = {
+    type: "object",
+    properties: {
+      start_date: {
+        type: "string",
+        format: "date-time",
+        title: "Start Date (Local Time)",
```
Comment (guynir42) [inline static/js/components/NewAPIObservation.jsx:117]: Just making sure: the dates given in local time are correctly translated to UTC in the DB?

**Review by guynir42 — COMMENTED** [review]
Minor comments.

File: static/js/components/NewAPIObservation.jsx (hunk 4 above)
Comment (mcoughlin) [reply static/js/components/NewAPIObservation.jsx:117]: @Theodlz How does this stuff work exactly? Where does the conversion to UTC happen and what do I need to do to ensure that occurs?

File: skyportal/facility_apis/ztf.py (hunk 2 above)
Comment (mcoughlin) [reply skyportal/facility_apis/ztf.py:445]: good point.

File: skyportal/handlers/api/observation.py (hunk 3 above)
Comment (mcoughlin) [reply skyportal/handlers/api/observation.py:680]: It's a bit of a funny name, but "retrieve" grabs executed observations from the API and commits them. Do you prefer a different name @guynir42 ?

File: skyportal/facility_apis/ztf.py (hunk 1 above)
Comment (mcoughlin) [reply skyportal/facility_apis/ztf.py]: fixed.

File: skyportal/handlers/api/observation.py (hunk 3 above)
Comment (guynir42) [reply skyportal/handlers/api/observation.py:680]: I am not sure if a different name is the thing. Maybe add a comment explaining what's happening.

Comment (mcoughlin) [reply skyportal/handlers/api/observation.py]: added a comment

File: static/js/components/NewAPIObservation.jsx (hunk 4 above)
Comment (mcoughlin) [reply static/js/components/NewAPIObservation.jsx:119]: Added the +00:00.

**Comment by mcoughlin** [comment]
@guynir42 Take another look?

File: skyportal/tests/flaky/test_frontend_flaky.py [hunk 5]
```diff
@@ -355,3 +355,147 @@ def test_gcnevents_observations(
...
+        token=super_admin_token,
+    )
+    assert status == 200
+    assert data['status'] == 'success'
+    instrument_id = data['data']['id']
+
+    # wait for the fields to populate
+    time.sleep(15)
+
+    print(super_admin_token)
```
Comment (guynir42) [inline skyportal/tests/flaky/test_frontend_flaky.py:416]: print statement.

File: skyportal/tests/flaky/test_frontend_flaky.py [hunk 6]
```diff
@@ -355,3 +355,147 @@ def test_gcnevents_observations(
...
+            'field_region': Regions.read(regionsdatafile).serialize(format='ds9'),
+        },
+        token=super_admin_token,
+    )
+    assert status == 200
+    assert data['status'] == 'success'
+    instrument_id = data['data']['id']
+
+    # wait for the fields to populate
+    time.sleep(15)
```
Comment (guynir42) [inline skyportal/tests/flaky/test_frontend_flaky.py:414]: can we replace this with a `wait_for_xpath`?

File: skyportal/tests/flaky/test_frontend_flaky.py [hunk 7]
```diff
@@ -355,3 +355,147 @@ def test_gcnevents_observations(
...
+        scroll_parent=True,
+    )
+
+    # Click somewhere outside to remove focus from instrument select
+    driver.click_xpath("//header")
+
+    driver.click_xpath(submit_button_xpath)
+
+    # wait for the observation plan to complete
+    time.sleep(15)
```
Comment (guynir42) [inline skyportal/tests/flaky/test_frontend_flaky.py:460]: same here.

**Review by guynir42 — COMMENTED** [review]
Minor comments on the test.

File: skyportal/tests/flaky/test_frontend_flaky.py (hunk 5 above)
Comment (mcoughlin) [reply skyportal/tests/flaky/test_frontend_flaky.py]: removed.

File: skyportal/tests/flaky/test_frontend_flaky.py (hunk 6 above)
Comment (mcoughlin) [reply skyportal/tests/flaky/test_frontend_flaky.py]: i settled on a loop Josh and I worked out for another one.

File: skyportal/tests/flaky/test_frontend_flaky.py (hunk 7 above)
Comment (mcoughlin) [reply skyportal/tests/flaky/test_frontend_flaky.py]: Gone.

**Comment by mcoughlin** [comment]
@guynir42 Permissions infrastructure not important here, consider merging if happier.

**Review by guynir42 — APPROVED** [review]

**Pull request merged**